File: sort/Queue.go
package sortPkg

type Node[T any] struct {
	Value T
	Next  *Node[T]
}

type Queue[T any] struct {
	Lenght int
	Head   *Node[T]
	Tail   *Node[T]
}

func (q *Queue[T]) Enqueue(item T) {
	node := &Node[T]{Value: item}
	q.Lenght++

	if q.Tail == nil {
		q.Head = node
		q.Tail = node
		return
	}

	q.Tail.Next = node
	q.Tail = node
}

func (q *Queue[T]) Deque() *T {
	if q.Head == nil {
		return nil
	}

	q.Lenght--

	Value := q.Head.Value
	q.Head = q.Head.Next
	if q.Head == nil {
		q.Tail = nil
	}
	return &Value
}

func (q *Queue[T]) Peek() T {
	return q.Head.Value
}

File: sort/Queue.ts
type Node<T> = {
    value: T
    next?: Node<T>
}
export default class Queue<T>{
    public lenght: number
    public head?: Node<T>
    public tail?: Node<T>

    constructor() {
        this.head = this.tail = undefined
        this.lenght = 0
    }

    enqueue(item?: T): void {
        const node = { value: item } as Node<T>
        this.lenght++
        if (!this.tail) {
            this.tail = this.head = node
            return
        }

        this.tail.next = node
        this.tail = node
    }

    deque(): T | undefined {
        if (!this.head) {
            return undefined
        }

        this.lenght--

        const head = this.head.value
        this.head = this.head.next
        if (this.head == undefined) {
            this.tail = undefined
        }

        return head
    }

    peek(): T | undefined { // returns the first element
        return this.head?.value
    }
}